cpp/src/arrow/array/data.cc

```diff
@@ -195,6 +195,7 @@ int GetNumBuffers(const DataType& type) {
     case Type::NA:
     case Type::STRUCT:
     case Type::FIXED_SIZE_LIST:
+    case Type::RUN_END_ENCODED:
       return 1;
```

Comment:
Why 1? Run-end encoded arrays have no top-level buffers (no null bitmap), do they? Or do we reserve an empty null bitmap slot?

Reply:
You're right. I believe it should be 0.

https://arrow.apache.org/docs/dev/format/Columnar.html#buffer-listing-for-each-layout

Reply:
Well, note that this function does not reflect the number of buffers in the IPC layout, but the number of buffers _in Arrow C++_.

We have special code to work around it in the IPC layer:
https://github.com/apache/arrow/blob/359f28ba9d62a5e8456d92dfbe5b16b790019edd/cpp/src/arrow/ipc/reader.cc#L330-L335
https://github.com/apache/arrow/blob/359f28ba9d62a5e8456d92dfbe5b16b790019edd/cpp/src/arrow/ipc/writer.cc#L157-L169

So we could choose to also have a dummy null bitmap entry (in Arrow C++) for run-end encoded arrays, or not.
@lidavidm @westonpace What are your opinions on this?

Reply:
Hmm, why do we have the dummy in the first place? 

I suppose either you have the dummy and you put the exceptions in IPC/C Data, or you don't have the dummy and you put the exceptions in things that assume the presence of a validity buffer. I would rather not have the dummy, but without looking at the code, maybe that introduces too many exceptions elsewhere.

Reply:
The dummy in null arrays might be for historical reasons, and was kept for compatibility in Arrow C++.
See https://issues.apache.org/jira/browse/ARROW-6379

Reply:
I think most of the code in Acero is used to the idea of a validity buffer not existing if the null count is 0.  So I think the easier path would be to have 0 buffers.  But I'm pretty sure I can work around it if we decide it's easier elsewhere to have one.

Reply:
I think REEs make thinking about nullability so different than normal that a dummy buffer won't really make things much easier. That's my impression without have much knowledge about all the codepaths that arrays go through.

Reply:
From looking at @zeroshade PR, I got reminded that the child `values` array will have a validity bitmap. That, of course doesn't map to logical validity of the REE elements, but allows for an `IsValid(i)` at the top-level array that can be implemented at the cost of binary searching for the physical index of `i` and checking the validity bitmap of the child `values` array. Maybe that can make integration easier and allows Acero to improve how it deals with REEs more incrementally?

Reply:
We shouldn't deviate from established practice on other datatypes.
* `IsValid` on `DictionaryArray` only queries the indices, not the dictionary values
* `IsValid` on `UnionArray` already returns false, even if the corresponding child entry is null

Also you'll notice that `IsValid` and `IsNull` are non-virtual methods.

Reply:
> Also you'll notice that IsValid and IsNull are non-virtual methods.

And making them `virtual` would have performance impacts on everything. This would be a different API just in case kernels need it. More of a `compute` concern than something to be part of the `ArrayData` interface.